### ISSUE TITLE
op leak detection works in a confusing way when they are disabled at test step level.

### ISSUE THREAD

**Issue by kt3k** [denoland/deno#21307]
Context: https://github.com/denoland/deno_std/issues/3512

If the sanitization is disabled at test step level like the below example:
```js
Deno.test("reproduce", async (t) => {
  await t.step({
    name: "reproduce",
    sanitizeOps: false,
    sanitizeResources: false,
    fn: () => {
      setTimeout(() => {}, 100);
    },
  });
});
```

The step succeeds by ignoring op leak, but the parent test case still detects the op leak and the test case fails.

By this design, when the user want to disable sanitization at a test step, they also need to disable it for all the parent steps/cases. This doesn't seem ideal situation.

Is it possible to ignore the op leaks at parent level when they are already ignored at child level?

**Comment by KyleJune** [comment]
If there were multiple test cases in the parent and the timeout resolved during one of the other test cases that sanitizeOps, I believe it would be that other test case that fails rather than the parent, since the parent would end without any leaking ops.

I don't think the behavior should change. That leaking op could potentially cause problems in other test cases that would be harder to detect without the sanitization. Maybe the error message could be updated to reflect the origin of the leaking op, which would make it easier to track down the source of the problematic op.

If we were to change behavior so other test cases ignore those leaking ops from a test case that had sanitization disabled, maybe that could be accomplished by doing a diff of the list of ops before and after the test, so that you can get the ids of all ops that should be ignored by other test cases that have sanitization. The downside to this would be if a leaking op is causing another test with sanitization to fail, it would be less obvious why.

Or we could do a middleground solution, ignore the ops from the test case with sanitization disabled unless the test case with sanitization fails. Then, if it fails, show both the error and the leaking op info just in case that leaking op is causing the failure.

**Comment by kt3k** [comment]
>  That leaking op could potentially cause problems in other test cases that would be harder to detect without the sanitization.

That's true, but I also feel the current situation defeats the meaning of `sanitizeOps: false` option.

I was able to reproduce the issue without test steps like the below:
```js
Deno.test("ignore op", { sanitizeOps: false, sanitizeResources: false }, () => {
  setTimeout(() => {}, 10);
});
for (const i of Array(20).keys()) {
  Deno.test("test with no leak " + i, async () => {
    await new Promise((resolve) => setTimeout(resolve, 1));
  });
}
```

This constantly fails at `test with no leak 4` on my machine (I guess it could be at different case in other machine/platform). Should I also set `{ sanitizeOps: false, sanitizeResources: false }` to `test with no leak 4` to pass this test case? Or should I move the first test case to the last? Both solutions feel very confusing to me.

**Comment by lucacasonato** [comment]
@kt3k Your suggestion implies that if any test in a file has `sanitizeOps` or `sanitizeResources` specified, all tests in that file should have their sanitizer disabled automatically. This seems bad.

We could just outright ban having a mix of sanitized and non-sanitized tests in the same scope? Either all tests are sanitized, or all tests unsanitized, or unsanitized steps inside a sanitized test/step.

**Comment by bartlomieju** [comment]
> We could just outright ban having a mix of sanitized and non-sanitized tests in the same scope? Either all tests are sanitized, or all tests unsanitized, or unsanitized steps inside a sanitized test/step.

I like this idea - should it be a separate API like: `Deno.test.disableSanitizer()`?

**Comment by lucacasonato** [comment]
@bartlomieju `Deno.test.disableSanitizersForFil()`?

@kt3k File

**Comment by bartlomieju** [comment]
That works for me. The only question I have is if we allow to disable certain sanitizers (ops/resources) or disable all of them. I think we should not allow to disable "exit sanitizer" at all. So maybe `Deno.test.disableSanitizersForFile({ ops: true, resources: true })` while if no args provided it disables both?
